docs/proposals/20-ha-control-planes.md

```diff
@@ -0,0 +1,2624 @@
+
+**Cons**
+* For a multi-zonal etcd setup since there is only one active etcd, all instances of API Server's spread across zones will connect to a single etcd instance. This results in a drastic increase in cross-zonal traffic, translating to higher operating cost.
+* As compared to an `active-active` etcd cluster there is not much difference in cost of compute resources (CPU, Memory, Storage)
+* etcd-druid will have to periodically check the health of both the `primary` and `hot-standby` nodes and ensure that these are up and running.
+* There will be a potential delay in determining that a `primary` etcd instance is no longer healthy. Thereby increasing the delay in switching to the `hot-standy` etcd instance causing longer downtime. It is also possible that at the same time `hot-standy` also went down or is otherwise unhealthy resulting in a complete down time. The amount of time it will take to recover from such a situation would be several minutes (time to start etcd pod + time to restore either from full snapshot or apply delta snapshots).
+* Synchronization is always via backup-bucket which will be less frequent as compared to an `active-active` etcd cluster where there is real-time synchronization done for any updates by the leader to majority or all of its followers. If the primary crashes, the time.
+* During the switchover from `primary` to `hot-standby` if the `hot-standy` etcd is in process of applying delta snaphots or restoring from a new full snapshot then `hot-standby` should ensure that the backup-restore container sets the readiness probe to indicate that it is not ready yet causing additional down time.
+</details>
+
+<details>
+<summary>Option-2: Independent single node clusters with Mirror-Maker</summary>
+<img src="assets/activepassive-option2.png">
+
+An extension to `Option-1` where we do not only rely on snapshots (full/delta) to keep the `hot-standy` etcd node in sync with the `primary etcd`. etcd provides an alternative way via [mirror maker](https://github.com/etcd-io/etcd/blob/main/etcdctl/doc/mirror_maker.md) to continuously relay key creates and updates to a separate cluster. In this option reconciliation from snapshots in backup-bucket will be used as a fallback.
+
+**Pros**
+* All the pros in `Option-1` are also applicable for this option
+* Using mirror maker provides continuous synchronization of key creates/updates to the `hot-standby` etcd, making it less dependent on reconciliaton from snapshots fetched from the backup-bucket.
+
+**Cons**
+* All the cons in `Option-1` are also applicable for this option.
+* Mirroring does not guarantee linearizability. During network disconnects, previous key-values might have been discarded, and clients are expected to verify watch responses for correct ordering. Thus, there is no ordering guarantee in mirror. So the client wishes to preserve all historical data and ordering then a learner approach of synchronization is better as compared to using a mirror maker.
```

Comment:
Is that a problem for Kubernetes API servers? Is `Mirror-Maker` used in this context at all?

Reply:
This option has been removed from the GEP as we have not evaluated this option and there is no point putting a theoretical option in this doc.